crates/task-impls/src/quorum_vote/mod.rs

```diff
@@ -285,17 +287,69 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES> + 'static, V: Versions> Handl
             match event.as_ref() {
                 #[allow(unused_assignments)]
                 HotShotEvent::QuorumProposalValidated(proposal, parent_leaf) => {
+                    let version = match self.upgrade_lock.version(self.view_number).await {
+                        Ok(version) => version,
+                        Err(e) => {
+                            tracing::error!("{e:#}");
+                            return;
+                        }
+                    };
+                    let mut is_same_block_proposed = false;
                     let proposal_payload_comm = proposal.data.block_header.payload_commitment();
-                    if let Some(ref comm) = payload_commitment {
+                    let parent_commitment = parent_leaf.commit(&self.upgrade_lock).await;
+                    let proposed_leaf = Leaf::from_quorum_proposal(&proposal.data);
+
+                    tracing::debug!(
+                        "proposal leaf height = {}, parent leaf height = {}",
+                        proposed_leaf.height(),
+                        parent_leaf.height()
+                    );
+                    if version >= V::Epochs::VERSION
+                        && self
+                            .consensus
+                            .read()
+                            .await
+                            .is_qc_forming_eqc(&proposal.data.justify_qc)
+                    {
+                        tracing::info!(
+                            "Reached end of epoch. Justify QC is for the last block in the epoch."
+                        );
+                        if proposed_leaf.height() != parent_leaf.height()
+                            || proposed_leaf.payload_commitment()
+                                != parent_leaf.payload_commitment()
+                        {
+                            tracing::error!("Justify QC is for the last block but it's not extended and a new block is proposed. Not voting!");
+                            return;
+                        }
+
+                        tracing::info!("Reached end of epoch. Proposed leaf has the same height and payload as its parent. No need to check the VID and DAC.");
+                        is_same_block_proposed = true;
+
+                        let consensus_reader = self.consensus.read().await;
+
+                        let Some(vid_shares) = consensus_reader
+                            .vid_shares()
+                            .get(&parent_leaf.view_number())
```

Comment:
Is this always getting the right VID share?  For the first proposal after last block this will of course be correct, but for the next one in the chain where do we set the VID share to equal the original block?

For example for the chain Proposal(bk, v) <- Proposal(bk, v+1) <- proposal(bk, v + 2) when do we set the VID share for v +1 (and v +2) equal to the one from v?